[Nice-to-have N1] Chat feature (and bug fixes in collaboration service frontend)

larrywang0701 opened the pull request:
# N1 Chat feature
Implemented N1 basic chat feature (both frontend and backend) #98 
# Bug fixes for my collaboration service frontend
 - Bug fixes and improvements to my collaboration service frontend as @brein62 mentioned in #106 
 - Also fixed a bug that the WebSocket will disconnect when the user switches view in the collaboration frontend


# Screenshots
![image](https://github.com/user-attachments/assets/14a483db-6540-40ed-904c-81f9ee9b5351)

![image](https://github.com/user-attachments/assets/c060aabb-ad0e-4ab4-b51b-4c65d3e9cf5e)

![image](https://github.com/user-attachments/assets/fd21ac5b-e6de-4e0b-9e25-e1317f8d9096)

![image](https://github.com/user-attachments/assets/160d4816-8e2b-4a8f-b3cf-6eb944fc664e)


(I have started working on the chat feature since many days ago, but I wait and open this PR at this time because I have to wait for PR #106 to get merged before opening this PR which is depended on #106)

brein62 commented:
as mentioned previously, split the **bugfixes** and the new **chat feature** into two separate PRs for us to properly review the code

brein62 commented on frontend/src/pages/collaboration-service/CollaborationPage.tsx:
```diff
@@ -115,21 +115,11 @@ export default function CollaborationPage() {
       <MainContainer className="px-4 text-center gap-3 flex flex-col">
         <h2 className="text-2xl">
-          You are at an invalid document ID for collaborating
+          Invalid collaboration arguments. Please try matching again.
```
better way to express this:

`It seems like you are not in a valid collaboration environment. Please try matching again.`

larrywang0701 replied:
done

larrywang0701 commented:
> as mentioned previously, split the **bugfixes** and the new **chat feature** into two separate PRs for us to properly review the code
> 
> an issue we faced when reviewing and reading your code is that i think you don't comment enough in your code, making it very difficult for us to read it. additionally, when we ask for your viewpoints and code explanation, sometimes you can MIA for days which makes things more difficult for us. your code is good, just that it's not communicated well to other developers (i mean you don't leave your fellow developers hanging in the workplace during your interns right!?)
> 
> i think the code so far here is fine and i can fully understand it, good job - just that you need to communicate more through your code since some of us aren't too well-versed in frontend

Seems that it's too late for you to say "split PRs" for now since the PR is already created and I already fixed bugs along with working on chat feature - so it may mess up the commits and code if I forcefully "separate" this PR into two PRs.

But these information may be helpful for you to review this PR in a way that separated chat feature and bug fixes in your mind:

Total 9 files changed in this PR:
  - The following changed files are for **bug fixes and improvements on my collaboration frontend**:
    - `frontend/src/components/collaboration-service/CodeEditingArea.tsx`
    - `frontend/src/components/collaboration-service/CodeEditorSettings.ts`
    - `frontend/src/components/collaboration-service/LayoutManager.tsx`
    - `frontend/src/components/collaboration-service/QuestionArea.tsx`
    - `frontend/src/components/common/VerticallySplitView.tsx` (renamed from `VerticallySplitedView.tsx`)
  - The following changed files are for **chat feature**:
    - `collaboration-service/server.ts`
    - `frontend/src/components/collaboration-service/ChatBubble.tsx` (new file)
    - `frontend/src/components/collaboration-service/ChattingOverlay.tsx` (new file)
  - The following changed files contains both chat feature and bug fixes:
    - `frontend/src/pages/collaboration-service/CollaborationPage.tsx`
 
Do reply me here if you still have any questions regarding my code, thanks!

brein62 commented:
> > as mentioned previously, split the **bugfixes** and the new **chat feature** into two separate PRs for us to properly review the code
> > an issue we faced when reviewing and reading your code is that i think you don't comment enough in your code, making it very difficult for us to read it. additionally, when we ask for your viewpoints and code explanation, sometimes you can MIA for days which makes things more difficult for us. your code is good, just that it's not communicated well to other developers (i mean you don't leave your fellow developers hanging in the workplace during your interns right!?)
> > i think the code so far here is fine and i can fully understand it, good job - just that you need to communicate more through your code since some of us aren't too well-versed in frontend
> 
> Seems that it's too late for you to say "split PRs" for now since the PR is already created and I already fixed bugs along with working on chat feature - so it may mess up the commits and code if I forcefully "separate" this PR into two PRs.
> 
> But these information may be helpful for you to review this PR in a way that separated chat feature and bug fixes in your mind:
> 
> Total 10 files changed in this PR:
> 
> * The following changed files are for **bug fixes and improvements on my collaboration frontend**:
>   
>   * `frontend/src/components/collaboration-service/CodeEditingArea.tsx`
>   * `frontend/src/components/collaboration-service/CodeEditorSettings.ts`
>   * `frontend/src/components/collaboration-service/LayoutManager.tsx`
>   * `frontend/src/components/collaboration-service/QuestionArea.tsx`
>   * `frontend/src/components/common/VerticallySplitView.tsx` (renamed from `VerticallySplitedView.tsx`)
>   * `frontend/src/contexts/CollaborationContext.tsx`
> * The following changed files are for **chat feature**:
>   
>   * `collaboration-service/server.ts`
>   * `frontend/src/components/collaboration-service/ChatBubble.tsx` (new file)
>   * `frontend/src/components/collaboration-service/ChattingOverlay.tsx` (new file)
> * The following changed files contains both chat feature and bug fixes:
>   
>   * `frontend/src/pages/collaboration-service/CollaborationPage.tsx`
> 
> Do reply me here if you still have any questions regarding my code, thanks!

looks good, thanks for the prompt reply! i will start testing and reviewing now thanks!

brein62 commented:
can't seem to get your PR to work

brein62 commented on frontend/src/pages/collaboration-service/CollaborationPage.tsx:
```diff
@@ -115,21 +115,11 @@ export default function CollaborationPage() {
       <MainContainer className="px-4 text-center gap-3 flex flex-col">
         <h2 className="text-2xl">
-          You are at an invalid document ID for collaborating
+          Invalid collaboration arguments. Please try matching again.
```
change this to `It seems like you are not in a valid collaboration environment. Please try matching again.` as well

larrywang0701 replied:
it's a duplicated `if` for checking null statements. I just deleted the whole deplicated `if` block

brein62 commented on frontend/src/components/collaboration-service/QuestionArea.tsx:
```diff
@@ -38,14 +44,25 @@ export default function QuestionArea({questionId} : {questionId : string}) {
   return (
     <>
-      <div className="ml-5">
+      <div className="ml-5 mr-5">
```
thanks for the fix

brein62 commented on frontend/src/components/collaboration-service/QuestionArea.tsx:
```diff
@@ -38,14 +44,25 @@ export default function QuestionArea({questionId} : {questionId : string}) {
           <div className="text-2xl font-bold">{question.title}</div>
           <hr className="mt-2 mb-2"/>
           <div className="text-sm">{question.description}</div>
+          <Button variant="destructive" className="bg-red-500 mt-16 hover:bg-red-300 ml-auto text-white" onClick={endSession}>End session</Button>
```
`<Button variant="destructive" className="bg-red-500 mt-16 hover:bg-red-300 ml-auto text-white" onClick={endSession}>End session</Button>`
->
`<Button variant="destructive" className="btnred mt-16 ml-auto text-white" onClick={endSession}>End session</Button>`

larrywang0701 replied:
OK

brein62 commented on frontend/src/pages/collaboration-service/CollaborationPage.tsx:
```diff
@@ -102,7 +102,7 @@ export default function CollaborationPage() {
+          It seems like you are not in a valid collaboration environment. Please try matching again.
         </h2>
         <div className="flex justify-center">
           <Button className="btnblack">
```
dont forget to add a PageHeader on top as well for all error messages that might occur

larrywang0701 replied:
OK

brein62 reviewed (changes requested):
thanks for the fixes! just a few changes. unable to reproduce results as my copy leads to all of

one of the below is null
CollaborationPage.tsx:98 roomId: null
CollaborationPage.tsx:99 matchedUser: null
CollaborationPage.tsx:100 question: null

resulting in invalid collab environment message when i try to match normally.

another frontend bug is that when you leave the page, the collab service text resets as well, but i'll try to get my side working first (if not, i'll propose some fixes)

brein62 commented:
another bug (since you're fixing bugs here): when you exit out to another page, the content resets as well.

Two ways to reproduce (but i think they're both the same bug):
1. Match User A with User B, resulting in collab page
2. User A clicks on "Practise an Interview"
3. Text for User A is blank now

1. Match User A with User B, resulting in collab page
2. User A clicks on "PeerPrep" logo on top left
3. User A goes back to collab page by clicking on "Practise an Interview"
4. Text for User A is blank now

larrywang0701 commented:
> another bug (since you're fixing bugs here): when you exit out to another page, the content for both the chat and the code window resets as well.
> 
> Two ways (but i think they're both the same bug):
> 
> First way
> 
> 1. Match User A with User B, resulting in collab page
> 2. User A clicks on "Practise an Interview"
> 3. Text for User A is blank now (code editor + message screen)
> 
> Second way
> 
> 1. Match User A with User B, resulting in collab page
> 2. User A clicks on "PeerPrep" logo on top left
> 3. User A goes back to collab page by clicking on "Practise an Interview"
> 4. Text for User A is blank now (code editor + message screen)
> 
> Functionality improvements:
> 
> * Add the username above the text for the message feature
> * Set minimum width of the message window to 300-400px since if you use half the screen width, the message window doesn't look too hot... (doesn't matter that it covers the code editor part)
> * Set maximum height of the editor + code pane to like 400px (or 100vh - 300px) so that the entire content fits within the page. for any overflows, it should be handled within the two frames, not the page. -> if u can't i can help with this
> 
> would still prefer changing the button for language to a dropdown but i can fix it if you dont know how (dont need to stress too much about it)

For the bug, I tried to fix by frontend side by adding a retry mechanism on loading (initial) code to editor, but I found that sometimes the frontend can correctly load code from the backend (after I navigate away and back or clicking Practise an Interview), but sometimes the bug still happens and from debug console I found that the code responsed from the backend is empty strings. Maybe this is a backend-side bug, so please let the people who are responsible for the collaboration backend to consider this bug from the backend, since I really don't know why frontend sometimes received empty code strings from `load-document` sent by the backend but sometimes not. Thanks!

For your functionality improvements, I've done (1)username and (2) minimum width of chat box, and for (3)maximum height for code editor I think 800px would be nicer (you can change to 400px by yourself later if you still think 400px is better, all personal thoughts about aesthetic view so no need to argue about this). For your "dropdown" things sorry I really cannot get your idea (I've already changed the language selection UI in this PR from my previous PR but you're still mentioning it, so I guess we are really thinking different things). Maybe you can do the "dropdown" you mean by yourself with your thoughts after merging this PR. Thanks!

larrywang0701 commented:
Can just merge if no further (new) problems

larrywang0701 commented:
To clarify why I think the bug of empty code editor may comes from backend:
![image](https://github.com/user-attachments/assets/dc1ac36b-2d3f-470c-b761-fdc8984e3f58)
When navigating back/reopening collaboration page, sometimes there will be a  "code: " output at browser's console. This means that `load-document` signal from the backend is received by the frontend through WebSocket, but with empty code string in the payload. So I think since the signal itself contains empty code string, and frontend received is successfully, I guess the bug may actually comes from the backend.

brein62 commented:
for my review my home internet is a bit on and off so i’ll need a while to review but hopefully can start working on my part by tonight

brein62 reviewed (approved):
LGTM! any remaining changes to be done in separate PRs

tiongMax reviewed (approved):
Lgtm!
